models/repo_list.go

```diff
@@ -114,51 +114,91 @@ type SearchRepoOptions struct {
 
 // SearchRepositoryByName takes keyword and part of repository name to search,
 // it returns results in given range and number of total results.
-func SearchRepositoryByName(opts *SearchRepoOptions) (repos RepositoryList, count int64, err error) {
-	var cond = builder.NewCond()
+func SearchRepositoryByName(opts *SearchRepoOptions) (repos RepositoryList, _ int64, _ error) {
+	// Set owner of searched repositories if present in options
+	searchOwnerID := opts.OwnerID
+
+	// Set owner ID to Searcher ID if Owner is not filled
+	if opts.OwnerID <= 0 && opts.Searcher != nil {
+		searchOwnerID = opts.Searcher.ID
+	}
+	// Check if user with Owner ID exists
+	if searchOwnerID > 0 {
+		userExists, err := GetUser(&User{ID: searchOwnerID})
+		if err != nil {
+			return nil, 0, err
+		}
+		if userExists == false {
+			return nil, 0, ErrUserNotExist{UID: searchOwnerID}
+		}
+	}
+
+	// Set public search by default
+	isPublicSearch := true
+
+	// Allow to search for owners private data
+	if opts.Searcher != nil && (opts.Searcher.ID == searchOwnerID || opts.Searcher.IsAdmin) {
```

Comment:
This may not be right; if the owner (i.e. `searchOwnerID`) is an organization, and the searcher is an owner of that organization, the searcher is allowed to view private repos, but with this implementation will not be able to.

More generally, I don't think we should be making authorization decisions in `models`; if the search should not return private repos, that is the responsibility of the caller (who should set `opts.Private` accordingly)